// next-intl/dist/development/server/react-client/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function notSupported(name) {
  return () => {
    throw new Error("`".concat(name, "` is not supported in Client Components.") );
  };
}

// Must match `../index.tsx`

// prettier-ignore
const getRequestConfig = () => notSupported('getRequestConfig');
// prettier-ignore
/** @deprecated Is called `getFormatter` now. */
const getIntl = notSupported('getIntl');
// prettier-ignore
const getFormatter = notSupported('getFormatter');
// prettier-ignore
/** @deprecated Please use the `locale` parameter from Next.js instead. */
const getLocale = notSupported('getLocale');
// prettier-ignore
const getNow = notSupported('getNow');
// prettier-ignore
const getTimeZone = notSupported('getTimeZone');
// prettier-ignore
/** @deprecated Is called `getTranslator` now. */
const getTranslations = notSupported('getTranslations');
// prettier-ignore
const getTranslator = notSupported('getTranslator');
// prettier-ignore
const getMessages = notSupported('getMessages');

exports.getFormatter = getFormatter;
exports.getIntl = getIntl;
exports.getLocale = getLocale;
exports.getMessages = getMessages;
exports.getNow = getNow;
exports.getRequestConfig = getRequestConfig;
exports.getTimeZone = getTimeZone;
exports.getTranslations = getTranslations;
exports.getTranslator = getTranslator;
