// next-intl/dist/development/navigation/react-server/createLocalizedPathnamesNavigation.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var getLocaleFromHeader = require('../../server/getLocaleFromHeader.js');
var utils = require('../utils.js');
var Link = require('../../link/react-server/Link.js');
var redirect = require('../../server/redirect.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

function createLocalizedPathnamesNavigation(_ref) {
  let {
    locales,
    pathnames
  } = _ref;
  function Link$1(_ref2) {
    let {
      href,
      locale,
      ...rest
    } = _ref2;
    const defaultLocale = getLocaleFromHeader.default();
    const finalLocale = locale || defaultLocale;
    return /*#__PURE__*/React__default.default.createElement(Link.default, _rollupPluginBabelHelpers.extends({
      href: utils.compileLocalizedPathname({
        locale: finalLocale,
        // @ts-expect-error -- This is ok
        pathname: href,
        // @ts-expect-error -- This is ok
        params: typeof href === 'object' ? href.params : undefined,
        pathnames
      }),
      locale: locale
    }, rest));
  }
  function redirect$1(href) {
    const locale = getLocaleFromHeader.default();
    const resolvedHref = utils.compileLocalizedPathname({
      ...utils.normalizeNameOrNameWithParams(href),
      locale,
      pathnames
    });
    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      args[_key - 1] = arguments[_key];
    }
    return redirect.default(resolvedHref, ...args);
  }
  function notSupported(message) {
    return () => {
      throw new Error("`".concat(message, "` is not supported in Server Components. You can use this hook if you convert the component to a Client Component."));
    };
  }
  return {
    Link: Link$1,
    redirect: redirect$1,
    usePathname: notSupported('usePathname'),
    useRouter: notSupported('useRouter')
  };
}

exports.default = createLocalizedPathnamesNavigation;
